### PR TITLE
Blocklistener

### PR DESCRIPTION
**Pull request by almogdepaz**
better name suggestions are wellcome

### CODE REVIEW

File: sync/block_listener.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+		exit:           make(chan struct{})}
+	bl.RegisterMsgHandler(BLOCK, newBlockRequestHandler(layers))
+	return &bl
+}
+
+func (bl *BlockListener) run() {
+	for {
+		select {
+		case <-bl.exit:
+			log.Debug("run stoped")
```
Comment (antonlerner) [inline sync/block_listener.go]: log.info

File: sync/block_listener.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+func NewBlockListener(net server.Service, bv BlockValidator, layers mesh.Mesh, timeout time.Duration, concurrency int) *BlockListener {
+	bl := BlockListener{
+		BlockValidator: bv,
+		Mesh:           layers,
+		Peers:          NewPeers(net),
+		MessageServer:  server.NewMsgServer(net, blockProtocol, timeout),
+		workers:        make(chan struct{}, concurrency),
+		unknownQueue:   make(chan mesh.BlockID, 200), //todo tune buffer size + get buffer from config
```
Comment (antonlerner) [inline sync/block_listener.go:50]: don't put in config, I'd put a const variable

File: sync/block_listener.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+func (bl *BlockListener) run() {
+	for {
+		select {
+		case <-bl.exit:
+			log.Debug("run stoped")
+			return
+		case id := <-bl.unknownQueue:
+			log.Debug("fetch block ", id, "buffer is at ", len(bl.unknownQueue)/cap(bl.unknownQueue), " capacity")
+			bl.workers <- struct{}{}
```
Comment (antonlerner) [inline sync/block_listener.go]: change name to workersSemaphore or something that better describes it

File: sync/block_listener.go (hunk 2 above)
Comment (antonlerner) [inline sync/block_listener.go:50]: rename to ingressQueue? why unknown?

File: sync/block_listener.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+			}()
+		}
+	}
+}
+
+//todo handle case where no peer knows the block
+func (bl *BlockListener) FetchBlock(id mesh.BlockID) {
+	for _, p := range bl.GetPeers() {
+		ch, err := sendBlockRequest(bl.MessageServer, p, id)
+		if err == nil {
```
Comment (antonlerner) [inline sync/block_listener.go]: refactor to get rid of indentaion levels to  
if err{
 break
}

File: sync/block_listener.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+			if b := <-ch; b != nil {
+				if bl.ValidateBlock(b) {
```
Comment (antonlerner) [inline sync/block_listener.go]: what happens if block is invalid

File: sync/block_listener.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+					bl.AddBlock(b)
+					//add all child blocks to unknown queue
+					bl.addUnknownToQueue(b)
```
Comment (antonlerner) [inline sync/block_listener.go]: if we received the block why continue iterating all other peers? - I'd break here

File: sync/block_listener.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+//todo handle case where no peer knows the block
+func (bl *BlockListener) FetchBlock(id mesh.BlockID) {
+	for _, p := range bl.GetPeers() {
+		ch, err := sendBlockRequest(bl.MessageServer, p, id)
+		if err == nil {
+			if b := <-ch; b != nil {
+				if bl.ValidateBlock(b) {
+					bl.AddBlock(b)
```
Comment (antonlerner) [inline sync/block_listener.go]: AddBlock needs to be thread safe - is it?

File: sync/block_listener.go (hunk 5 above)
Comment (almogdepaz) [reply sync/block_listener.go]: on to the next one, we don't keep syntactically invalid blocks

File: sync/syncer_test.go [hunk 8]
```diff
@@ -72,7 +72,7 @@ func TestSyncer_Start(t *testing.T) {
 	syncs, _ := SyncMockFactory(2, conf, "TestSyncer_Start_")
```
Comment (antonlerner) [inline sync/syncer_test.go:72]: Id like for us to write random tests, if you want you can define what a random UT would look like and we'll let a contributor do it

**Review by antonlerner — APPROVED** [review]

File: sync/block_listener.go (hunk 6 above)
Comment (almogdepaz) [reply sync/block_listener.go]: yup, missed it.

File: sync/block_listener.go (hunk 7 above)
Comment (almogdepaz) [reply sync/block_listener.go]: it is

**Pull request merged**